torchx/test/fixtures.py

```diff
@@ -0,0 +1,167 @@
+        and writes the given content line-by-line into the file. Returns the filepath to the written file.
+
+        Usage:
+
+        .. code-block:: python
+
+         self.write("foo.txt", content=["hello", "world"])
+         self.write("foo/bar.txt", content=["hello", "world"])
+
+        """
+
+        f = self.touch(filepath)
+        with open(f, "w") as fout:
+            fout.writelines(content)
+        return f
+
+    def read(self, filepath: Union[str, Path]) -> List[str]:
+        """
+        Reads the entire contents of the file into a list.
+        The ``filepath`` is assumed to be relative to the test tmpdir
+
+
+        Usage:
+
+        .. code-block:: python
+
+         f = self.write("foo/bar.txt", content=["hello", "world"])
+         self.read(f) == ["hello", "world"]
+
+        """
+
+        with open(self.tmpdir / filepath, "r") as fin:
+            return fin.readlines()
+
+
+Ret = TypeVar("Ret")
+
+
+class DistributedTestCase(TestWithTmpDir):
+    """
+    A ``unittest.TestCase`` that has utility methods to run tests that need to be run in the context
+    of ``torch.distributed``.
+
+    Usage:
+
+    .. doctest::
+
+        >>> from torchx.test.fixtures import DistributedTestCase
+        >>> import torch.distributed as dist
+
+        >>> class MyDistributedTest(DistributedTestCase):
+        ...     @staticmethod
+        ...     def run_test(arg1) -> str:
+        ...         dist.init_process_group(backend="gloo")
+        ...         # run whatever needs to be tested
+        ...         return f"rank={dist.get_rank()}/{dist.get_world_size()} arg1={arg1}"
+        ...     def test_foo(self) -> None:
+        ...         ret = self.run_ddp(world_size=2, fn=MyDistributedTest.run_test)("hello-world")
+        ...         self.assertDictEqual(
+        ...           {
+        ...             0: "rank=0/2 arg1=hello-world",
+        ...             1: "rank=1/2 arg1=hello-world",
+        ...           },
+        ...           ret
+        ...         )
+        >>> MyDistributedTest().test_foo()
+    """
+
+    def run_ddp(
+        self, world_size: int, fn: Callable[..., Ret]
+    ) -> Callable[..., Dict[int, Ret]]:
+        """
+        Runs ``world_size`` copies of ``fn`` (one on each sub-process) as a DDP job on the local host.
+
+        .. note::
+            You MUST initialize the default process group as ``ape.distributed.util.init_process_group()``
+            in your ``fn`` before running any distributed/collective operations.
+
+        See class docstring for usage example.
+        """
+        config = LaunchConfig(
+            min_nodes=1,
+            max_nodes=1,
+            nproc_per_node=world_size,
+            rdzv_backend="c10d",
+            rdzv_endpoint="localhost:0",
+            max_restarts=0,
+            monitor_interval=0.01,
+            log_dir=str(self.tmpdir),
+            tee=Std.ALL,
+        )
+
+        return elastic_launch(config, entrypoint=fn)
```

Comment:
maybe we should provide a torchx method to do this w/ pickle

Reply:
so this actually launches the function (using torch.multiprocessing) under the hood. Its not playing the "write the function into a script" trick.